cpp/src/distance/cluster_cost.cuh

```diff
@@ -0,0 +1,79 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <raft/cluster/kmeans.cuh>
+#include <raft/distance/distance_types.hpp>
+#include <raft/distance/fused_l2_nn.cuh>
+#include <raft/handle.hpp>
+
+namespace raft::cluster::kmeans::runtime {
+template <typename ElementType, typename IndexType>
+void cluster_cost(const raft::handle_t& handle,
+                  const ElementType* X,
+                  IndexType n_samples,
+                  IndexType n_features,
+                  IndexType n_clusters,
+                  const ElementType* centroids,
+                  ElementType* cost)
+{
+  rmm::device_uvector<char> workspace(n_samples * sizeof(IndexType), handle.get_stream());
+
+  rmm::device_uvector<ElementType> x_norms(n_samples, handle.get_stream());
+  rmm::device_uvector<ElementType> centroid_norms(n_clusters, handle.get_stream());
+  raft::linalg::rowNorm(
+    x_norms.data(), X, n_features, n_samples, raft::linalg::L2Norm, true, handle.get_stream());
+  raft::linalg::rowNorm(centroid_norms.data(),
+                        centroids,
+                        n_features,
+                        n_clusters,
+                        raft::linalg::L2Norm,
+                        true,
+                        handle.get_stream());
+
+  auto min_cluster_distance =
+    raft::make_device_vector<raft::KeyValuePair<IndexType, ElementType>>(handle, n_samples);
+  raft::distance::fusedL2NNMinReduce(min_cluster_distance.data_handle(),
+                                     X,
+                                     centroids,
+                                     x_norms.data(),
+                                     centroid_norms.data(),
+                                     n_samples,
+                                     n_clusters,
+                                     n_features,
+                                     (void*)workspace.data(),
+                                     true,
+                                     true,
+                                     handle.get_stream());
+
+  auto distances = raft::make_device_vector<ElementType, IndexType>(handle, n_samples);
+  thrust::transform(
+    handle.get_thrust_policy(),
+    min_cluster_distance.data_handle(),
+    min_cluster_distance.data_handle() + n_samples,
+    distances.data_handle(),
+    [] __device__(const raft::KeyValuePair<IndexType, ElementType>& a) { return a.value; });
+
+  rmm::device_scalar<ElementType> device_cost(0, handle.get_stream());
+  raft::cluster::kmeans::cluster_cost(
```

Comment:
I think we're doing an additional computation here that can be avoided by running the `fusedL2NNMinReduce` w/ `sqrt=False`. If we have the squared distances already then we should just be able sum them up to get the `inertia` score.

Reply:
Could do a simple L1 norm computation for the reduction, I guess. https://github.com/rapidsai/raft/blob/d6df557eeb8388c74bc0135c0a32681f887b8bb8/cpp/include/raft/linalg/norm.cuh#L109

Reply:
thats a good point - and I've made the change to fusedL2NNMinReduce w/ sqrt=False.   I checked the cluster cost, and its already only doing a sum reduction - so using sqrt=True was wrong here.  

This also lets us avoid the issue in #1036 - afaict you are right and this issue is only with sqrt=True (all the tests pass at least now =) ).